### PR TITLE
Feat/batch set weights

### LINKED ISSUE #1018: Add batched extrinsics for setting weights
Issue: The utility pallet batching extrinsics require users to pay fees proportionally to the number of batched extrinsics, despite the `set_weights` being free. There's also chain bloating issue for multiple `set_weights` being called.

Solution: Add methods `batch_set_weights` and `batch_commit_weights` that would allow to set/commit weights to the multiple subnets from one hotkey using one extrinsic.

AC: 

- [x] Extrinsics
- [x] unit tests for happy path
- [x] unit tests for failures  
- [x] e2e tests for happy path
- [ ] e2e tests for rate limiting

### PR DESCRIPTION
**Pull request by camfairchild**
## Description

Fixes an issue with batching weights having an increased fee from: https://github.com/paritytech/polkadot-sdk/pull/5644

This adds both `batch_set_weights` and `batch_commit_weights`

Support added to bittensor in https://github.com/opentensor/bittensor/pull/2485

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Other (please describe):

## Breaking Change

If this PR introduces a breaking change, please provide a detailed description of the impact and the migration path for existing applications.

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have run `cargo fmt` and `cargo clippy` to ensure my code is formatted and linted correctly
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

## Screenshots (if applicable)

Please include any relevant screenshots or GIFs that demonstrate the changes made.

## Additional Notes

Please provide any additional information or context that may be helpful for reviewers.

### CODE REVIEW

File: pallets/subtensor/src/subnets/weights.rs [hunk 1]
```diff
@@ -105,6 +106,77 @@ impl<T: Config> Pallet<T> {
...
+        log::debug!(
+            "do_batch_commit_weights( origin:{:?} netuids:{:?}, hashes:{:?}",
+            hotkey,
+            netuids,
+            commit_hashes
+        );
+
+        let results: Vec<dispatch::DispatchResult> = netuids
+            .iter()
+            .zip(commit_hashes.iter())
```
Comment (gztensor) [inline pallets/subtensor/src/subnets/weights.rs:159]: If commit_hashes vector is longer than netuids vector, we will be committing weights for root, which will fail. Is this intentional that we don't check for vector lengths to match?

Comment (camfairchild) [reply pallets/subtensor/src/subnets/weights.rs:159]: added a check

**Review by gztensor — DISMISSED** [review]

**Review by gztensor — DISMISSED** [review]

**Review by sam0x17 — DISMISSED** [review]

File: pallets/subtensor/src/tests/batch_tx.rs [hunk 2]
```diff
@@ -33,3 +40,360 @@ fn test_batch_txs() {
...
+        step_block(7200 + 1);
+
+        // Set weights on the other hotkey and Use maximum value for u16
+        let weights: Vec<(Compact<u16>, Compact<u16>)> = vec![(Compact(1), Compact(u16::MAX))];
+        let version_key_0: Compact<u64> = SubtensorModule::get_weights_version_key(netuid_0).into();
+        let version_key_1: Compact<u64> = SubtensorModule::get_weights_version_key(netuid_1).into();
+        let version_key_2: Compact<u64> = SubtensorModule::get_weights_version_key(netuid_2).into();
+
+        // Set the min stake very high
+        SubtensorModule::set_weights_min_stake(stake_to_give_child * 5);
```
Comment (camfairchild) [inline pallets/subtensor/src/tests/batch_tx.rs]:
```suggestion
        SubtensorModule::set_stake_threshold(stake_to_give_child * 5);
```

File: pallets/subtensor/src/tests/batch_tx.rs [hunk 3]
```diff
@@ -33,3 +40,360 @@ fn test_batch_txs() {
...
+
+        // Check the key has less stake than required
+        assert!(
+            SubtensorModule::get_stake_for_hotkey_on_subnet(&hotkey, netuid_0)
+                < SubtensorModule::get_weights_min_stake()
```
Comment (camfairchild) [inline pallets/subtensor/src/tests/batch_tx.rs]:
```suggestion
                < SubtensorModule::get_stake_threshold()
```

File: pallets/subtensor/src/tests/batch_tx.rs [hunk 4]
```diff
@@ -33,3 +40,360 @@ fn test_batch_txs() {
...
+            System::events()
+        );
+
+        // Reset the events
+        System::reset_events();
+
+        assert!(!SubtensorModule::check_weights_min_stake(&hotkey, netuid_0));
+
+        // Set a minimum stake to set weights
+        SubtensorModule::set_weights_min_stake(stake_to_give_child - 5);
```
Comment (camfairchild) [inline pallets/subtensor/src/tests/batch_tx.rs]:
```suggestion
        SubtensorModule::set_stake_threshold(stake_to_give_child - 5);
```

File: pallets/subtensor/src/tests/batch_tx.rs [hunk 5]
```diff
@@ -33,3 +40,360 @@ fn test_batch_txs() {
...
+
+        // Check if the stake for the hotkey is above
+        assert!(
+            SubtensorModule::get_stake_for_hotkey_on_subnet(&hotkey, netuid_0)
+                >= SubtensorModule::get_weights_min_stake()
```
Comment (camfairchild) [inline pallets/subtensor/src/tests/batch_tx.rs]:
```suggestion
                >= SubtensorModule::get_stake_threshold()
```

File: pallets/subtensor/src/tests/batch_tx.rs [hunk 6]
```diff
@@ -33,3 +40,360 @@ fn test_batch_txs() {
...
+        // Reset the events
+        System::reset_events();
+
+        // Enable commit reveal for all networks
+        SubtensorModule::set_commit_reveal_weights_enabled(netuid_0, true);
+        SubtensorModule::set_commit_reveal_weights_enabled(netuid_1, true);
+        SubtensorModule::set_commit_reveal_weights_enabled(netuid_2, true);
+
+        // Set a minimum stake to set weights
+        SubtensorModule::set_weights_min_stake(stake_to_give_child - 5);
```
Comment (camfairchild) [inline pallets/subtensor/src/tests/batch_tx.rs]:
```suggestion
        SubtensorModule::set_stake_threshold(stake_to_give_child - 5);
```

File: pallets/subtensor/src/tests/batch_tx.rs [hunk 7]
```diff
@@ -33,3 +40,360 @@ fn test_batch_txs() {
...
+
+        // Check if the stake for the hotkey is above
+        assert!(
+            SubtensorModule::get_stake_for_hotkey_on_subnet(&hotkey, netuid_0)
+                >= SubtensorModule::get_weights_min_stake()
```
Comment (camfairchild) [inline pallets/subtensor/src/tests/batch_tx.rs]:
```suggestion
                >= SubtensorModule::get_stake_threshold()
```

**Review by sam0x17 — APPROVED** [review]

**Pull request merged**